proposals/fast-lane-ipns.md

```diff
@@ -0,0 +1,336 @@
+-->
+<!--
+For ease of discussion in PRs, consider breaking lines after every sentence or long phrase.
+-->
+
+## Purpose &amp; impact 
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+##### Problem statement
+
+Today [IPNS][] fails to provide reliable mutability primitive. There is no implementation that works across go, node and web, which makes it impractical.
+
+| Routing | web  | go   | nodejs |
+| ------- | ---- | ---- | ------ |
+| DHT     | ❌    | ✅    | ❌      |
+| pubsub  | ⚠️    | ⚠️    | ⚠️      |
+
+> ❌ - Not available
+> ✅ - Available
+> ⚠️ - Not out of the box
+>
+> - ❌ DHT in web nodes is impractical as most nodes will be undialable and so will be a web node.
+> - ❌ DHT implementation in nodejs is buggy, untersted and unoptimized ([ipfs/js-ipfs#3469][]). 
+> - ⚠️ PubSub in web nodes only works in swarms formed around central ICE server that are disjoint from go nodes. In practice creating fragmented networks of ephemeral peers.
+> - ⚠️ IPNS over pubsub is disabled in go by default (Can be enabled by running daemon with `--enable-namesys-pubsub`). Pubusb itself is also disabled by default ([protocol/web3-dev-team#53][]).
+> - ⚠️ IPNS over pubsub in nodejs still requires DHT to bootstrap overlay network of participating peers. Lack of proper DHT implementation prevents it. Which could be overcome via delegated routing but it does not work out of the box.
+
+Only DHT based implementation in go works out of the box. But it still fails to meet user expectations because:
+  1. *Slow* publish and resolution.
+  2. Requires continues republishing. (see [ipfs/go-ipfs#4435])
+
+Teams that have tried using IPNS in their products either switched to using [DNSLink][] or opted-into IPNS over pubsub and still find it *(too slow)* for certain updates *(Textile)*.
+
+To meet a market fit for products with changing state requirement _(that is anything interactive)_ IPNS needs to work reliably, fast _(comparable to web2 solutions)_ and across all supported enviroments _(go, web, nodejs)_.
+
+##### Proposition
+
+At the high level this proposal promises an efficienecy of centralized publishing and resolution while retaining resilience of distributed system.
+
+Proposed name resolution system is inspired by Domain Name System (DNS) as it introduces athoritative name resolvers _(from now on referred as)_ **name keeper** nodes. But unlike DNS, **name owners** _(private key holders)_ are in charge of choosing a name keepers and retain ability to change a name keeper in the future.
```

Comment:
How does this actually work? We could add a set of "keepers" to the IPNS record itself, I guess (as long as it's under a signature). I assume that's what you mean, right?

Reply:
I think I go into much greater detail below, but general idea is that instead of having to route on every name resolution, we'd do it very infrequently because there will be specific authoritative node you can ask about current record. That way nodes will be able to resolve "name keeper" address once and cache for future name resolutions.

Hypothesis is that while name keeper can change, it will be infrequent enough that we can see huge gains.

Reply:
Ah, sorry, my question is: how are you specifying these "name keepers"? Where does the list live, how is it published, how is it authenticated, how do I revoke keepers, etc.

Reply:
I think you may have meant how nome owners choose name keepers. I don't have thought through the specific here, but general idea is that name owner would provide a name keeper with a signed certificate, which it can use to prove to routing nodes that it is indeed the name keeping authority. 

That said I'm pretty open to discussing / reconsidering many details  of this proposal, mostly I'm pitching a bigger picture, that is lets make name publishing fast point to point operation. And make name resolution as close to that as possible.

Reply:
Got it. My thinking is that the keepers should be added as additional fields to the IPNS records. That way, there's no way to "roll back".

Reply:
> how are you specifying these "name keepers"?

1. You can create DNSLink record with `/ipns/${key}` and multiaddrs to name keeper.
2. You issue a certificate to a name keeper authorizing them as a name keeper. Which they use to let routing nodes know & proof they are the authority for the name. General assumption here is that pinning services will assume "name keeper" role as they are always on, well connected and you tell them what to pin anyway.

> Where does the list live

"Initial resolution optimization" attempts to describe it at high level. It assumes that infrastructure nodes can keep a list of "name keepers" and respond to queries from the network. If the name is unknown they can fallback to regular resolution DHT/PubSub and then cache the record forever. Note that here they resolve a "name keeper" not an actual record, and because those likely to change rarely no significant load is expected since there multiple layers of long term caching.

> how is it published, how is it authenticated

Name keepers are in charge of publishing to DHT, and over pubsub (possibly rendezvous) announcing themselves as "name keeper" for a name. Again since name keepers are unlikely to change TTLs could be very high.

> how do I revoke keepers, etc.

Again have not thought through all the details, but broadly speaking "name owner" would issue new certificate overriding old one to a different "name owner", which will than be responsible for updating the network. I expect pubsub would be an effective way to let routing routing nodes about the name keeper change. That said I'm open to alternative here as well.

Reply:
I realize I'm pretty vague when I talk about certificates, that is because I intentionally have not spend time thinking through all the details until this gets some traction. That said I think [UCAN](https://blog.fission.codes/auth-without-backend/)s are providing a pretty good and generalized model.